pkg/gofr/datasource/opentsdb/put.go

```diff
@@ -0,0 +1,354 @@
+	}
+
+	datapointGroups := make([][]DataPoint, 0)
+	datapointGroups = c.appendDataPoints(datasBytes, datapoints, datapointGroups)
+	status = StatusSuccess
+	message = "splitting into groups successful"
+
+	return datapointGroups, nil
+}
+
+func (c *Client) appendDataPoints(datasBytes []byte, datapoints []DataPoint, datapointGroups [][]DataPoint) [][]DataPoint {
+	if len(datasBytes) > c.opentsdbCfg.MaxContentLength {
+		return c.splitLargeDataPoints(datapoints, datapointGroups)
+	}
+
+	return append(datapointGroups, datapoints)
+}
+
+func (c *Client) splitLargeDataPoints(datapoints []DataPoint, datapointGroups [][]DataPoint) [][]DataPoint {
+	datapointsSize := len(datapoints)
+	startIndex := 0
+	endIndex := c.calculateEndIndex(datapointsSize)
+
+	for endIndex <= datapointsSize {
+		tempdps := datapoints[startIndex:endIndex]
+		if c.canAppendGroup(tempdps) {
+			datapointGroups = append(datapointGroups, tempdps)
+			startIndex = endIndex
+			endIndex = c.calculateNextEndIndex(startIndex, datapointsSize, len(tempdps))
+		} else {
+			endIndex -= c.opentsdbCfg.DetectDeltaNum
+		}
+
+		if startIndex >= datapointsSize {
+			break
+		}
+	}
+
+	return datapointGroups
+}
+
+func (c *Client) calculateEndIndex(datapointsSize int) int {
+	if datapointsSize > c.opentsdbCfg.MaxPutPointsNum {
+		return c.opentsdbCfg.MaxPutPointsNum
+	}
+
+	return datapointsSize
+}
+
+func (*Client) calculateNextEndIndex(startIndex, datapointsSize, tempSize int) int {
+	endIndex := startIndex + tempSize
+	if endIndex > datapointsSize {
+		return datapointsSize
+	}
+
+	return endIndex
+}
+
+func (c *Client) canAppendGroup(datapoints []DataPoint) bool {
+	tempdpsBytes, _ := json.Marshal(&datapoints)
+	return len(tempdpsBytes) <= c.opentsdbCfg.MaxContentLength
+}
+
+func parsePutErrorMsg(resp *PutResponse) error {
+	buf := bytes.Buffer{}
+	buf.WriteString(fmt.Sprintf("Failed to put %d datapoint(s) into opentsdb, statuscode %d:\n", resp.Failed, resp.StatusCode))
+
+	if len(resp.Errors) > 0 {
+		for _, putError := range resp.Errors {
+			buf.WriteString(fmt.Sprintf("\t%s\n", putError.String()))
+		}
+	}
+
+	return errors.New(buf.String())
+}
+
+func getPutBodyContents(datas []DataPoint) (string, error) {
+	if len(datas) == 1 {
+		result, err := json.Marshal(datas[0])
+		if err != nil {
+			return "", fmt.Errorf("failed to marshal datapoint: %v", err)
+		}
+
+		return string(result), nil
+	}
+
+	reqBodyCnt, err := marshalDataPoints(datas)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal datapoint: %v", err)
+	}
+
+	return reqBodyCnt, nil
+}
+
+func marshalDataPoints(datas []DataPoint) (string, error) {
+	buffer := bytes.NewBuffer(nil)
+	size := len(datas)
+
+	buffer.WriteString("[")
+
+	for index, item := range datas {
+		result, err := json.Marshal(item)
+		if err != nil {
+			return "", err
+		}
+
+		buffer.Write(result)
+
+		if index < size-1 {
+			buffer.WriteString(",")
+		}
+	}
+
+	buffer.WriteString("]")
+
+	return buffer.String(), nil
+}
```

Comment:
I don't get why you are not using

json.Marshall(datas)

A list of DataPoint will be serialized the way you expect no?

Unless, if you have a nil list, but here you should handle a nil and assign an empty list